src/float/conv.rs

```diff
@@ -9,70 +9,180 @@ use super::Float;
 /// These are hand-optimized bit twiddling code,
 /// which unfortunately isn't the easiest kind of code to read.
 ///
-/// The algorithm is explained here: <https://blog.m-ou.se/floats/>
+/// The algorithm is explained here: <https://blog.m-ou.se/floats/>. It roughly does the following:
+/// - Calculate the exponent based on the base-2 logarithm of `i` (leading zeros)
+/// - Calculate a base mantissa by shifting the integer into mantissa position
+/// - Figure out if rounding needs to occour by classifying truncated bits. Some patterns apply
+///   here, so they may be "squashed" into smaller numbers to spmplifiy the classification.
 mod int_to_float {
+    use super::*;
+
+    /// Calculate the exponent from the number of leading zeros.
+    fn exp<I: Int, F: Float<Int: CastFrom<u32>>>(n: u32) -> F::Int {
+        F::Int::cast_from(I::BITS + F::EXPONENT_BIAS - 2 - n)
+    }
+
+    /// Shift the integer into the float's mantissa bits. Keep the lowest exponent bit intact.
+    fn m_base<I: Int, F: Float<Int: CastFrom<I>>>(i_m: I) -> F::Int {
+        F::Int::cast_from(i_m >> ((I::BITS - F::BITS) + F::EXPONENT_BITS))
+    }
+
+    /// Calculate the mantissa in cases where the float size is greater than integer size
+    fn m_f_gt_i<I: Int, F: Float<Int: CastFrom<I>>>(i: I, n: u32) -> F::Int {
+        F::Int::cast_from(i) << (F::SIGNIFICAND_BITS - I::BITS + 1 + n)
+    }
+
+    /// Calculate the mantissa and a dropped bit adjustment  when `f` and `i` are equal sizes
+    fn m_f_eq_i<I: Int + CastInto<F::Int>, F: Float<Int = I>>(i: I, n: u32) -> (F::Int, F::Int) {
+        let base = (i << n) >> F::EXPONENT_BITS;
+
+        // Only the lowest `F::EXPONENT_BITS` bits will be truncated. Shift them
+        // to the highest position
+        let adj = (i << n) << (F::SIGNIFICAND_BITS + 1);
+
+        (base, adj)
+    }
+
+    /// Adjust a mantissa with dropped bits
+    fn m_adj<F: Float>(m_base: F::Int, dropped_bits: F::Int) -> F::Int {
+        // Branchlessly extract a `1` if rounding up should happen
+        let adj = (dropped_bits - (dropped_bits >> (F::BITS - 1) & !m_base)) >> (F::BITS - 1);
+
+        // Add one when we need to round up. Break ties to even.
+        m_base + adj
+    }
+
+    /// Combine a final float repr from an exponent and mantissa.
```

Comment:
It's an important detail that it doesn't just combine the exponent and mantissa fields, but instead is designed to take an off-by-one exponent and a mantissa with an extra 1 bit on the high side.